README.rst
.. image:: https://github.com/koenvervloesem/bluetooth-clocks/workflows/tests/badge.svg
    :alt: Continuous Integration
    :target: https://github.com/koenvervloesem/bluetooth-clocks/actions
.. image:: https://img.shields.io/pypi/v/bluetooth-clocks.svg
    :alt: Python package version
    :target: https://pypi.org/project/bluetooth-clocks/
.. image:: https://img.shields.io/pypi/pyversions/bluetooth-clocks.svg
    :alt: Supported Python versions
    :target: https://python.org/
.. image:: https://readthedocs.org/projects/bluetooth-clocks/badge/?version=latest
    :target: https://bluetooth-clocks.readthedocs.io/en/latest/?badge=latest
    :alt: Documentation Status
.. image:: https://codecov.io/gh/koenvervloesem/bluetooth-clocks/branch/main/graph/badge.svg?token=RQNLC3OTFN
    :alt: Codecov coverage
    :target: https://codecov.io/gh/koenvervloesem/bluetooth-clocks
.. image:: https://img.shields.io/github/license/koenvervloesem/bluetooth-clocks.svg
    :alt: License
    :target: https://github.com/koenvervloesem/bluetooth-clocks/blob/main/LICENSE.txt

|

================
Bluetooth Clocks
================


    Set and get the time on various Bluetooth Low Energy clocks


This project offers a way to easily recognize Bluetooth Low Energy (BLE) clocks from
their advertisements and has a device-independent API to set and get the time on them.

.. image:: https://github.com/koenvervloesem/bluetooth-clocks/raw/main/docs/_static/synchronized-clocks.jpg
    :alt: Synchronize all your Bluetooth Low Energy clocks

.. inclusion-marker-after-intro

Supported devices
=================

Bluetooth Clocks supports the following devices:

+--------------------------+------------+-------------------+-----------+
| Device                   | Set time   | Set 12/24h format | Read time |
+==========================+============+===================+===========+
| `Current Time Service`_  | Yes        | No                | Yes       |
| (e.g. PineTime with      |            |                   |           |
| InfiniTime firmware)     |            |                   |           |
+--------------------------+------------+-------------------+-----------+
| `PVVX firmware`_         | Yes        | No                | Yes       |
| (LYWSD03MMC, MHO-C401,   |            |                   |           |
| CGG1, CGDK2, MJWSD05MMC, |            |                   |           |
| MHO-C122)                |            |                   |           |
+--------------------------+------------+-------------------+-----------+
| Qingping BT Clock Lite   | Yes        | No                | No        |
+--------------------------+------------+-------------------+-----------+
| ThermoPro TP358/TP393    | Yes        | Yes               | No        |
+--------------------------+------------+-------------------+-----------+
| Xiaomi LYWSD02           | Yes        | No                | Yes       |
+--------------------------+------------+-------------------+-----------+

.. _Current Time Service: https://www.bluetooth.com/specifications/specs/current-time-service-1-1/
.. _PVVX firmware: https://github.com/pvvx/ATC_MiThermometer

.. inclusion-marker-before-installation

Installation
============

You can install bluetooth-clocks as a package from PyPI with pip::

    pip install bluetooth-clocks

Usage of the command-line program
=================================

If you have installed the package with ``pip``, you can run the program as ``bluetooth-clocks``::

    $ bluetooth-clocks -h
    usage: bluetooth-clocks [-h] [--version] [-v] [-vv] {discover,get,set} ...

    Bluetooth Clocks

    options:
      -h, --help           show this help message and exit
      --version            show program's version number and exit
      -v, --verbose        set loglevel to INFO
      -vv, --very-verbose  set loglevel to DEBUG

    Subcommands:
      {discover,get,set}
        discover           discover supported Bluetooth clocks
        get                get the time from a Bluetooth clock
        set                set the time of a Bluetooth clock

Discovering Bluetooth clocks
----------------------------

You can discover supported Bluetooth clocks with ``bluetooth-clocks discover``::

    $ bluetooth-clocks discover
    Scanning for supported clocks...
    Found a ThermoPro TP358: address BC:C7:DA:6A:52:C6, name TP358 (52C6)
    Found a Xiaomi LYWSD02: address E7:2E:00:B1:38:96, name LYWSD02
    Found a ThermoPro TP393: address 10:76:36:14:2A:3D, name TP393 (2A3D)
    Found a Qingping BT Clock Lite: address 58:2D:34:54:2D:2C, name Qingping BT Clock Lite
    Found a Current Time Service: address EB:76:55:B9:56:18, name F15

These are the options that the ``discover`` subcommand recognizes::

    $ bluetooth-clocks discover -h
    usage: bluetooth-clocks discover [-h] [-s SCAN_DURATION]

    options:
      -h, --help            show this help message and exit
      -s SCAN_DURATION, --scan-duration SCAN_DURATION
                            scan duration (default: 5 seconds)

Setting the time
----------------

Set the time of a clock with a given Bluetooth address::

    $ bluetooth-clocks set -a E7:2E:00:B1:38:96
    Scanning for device E7:2E:00:B1:38:96...
    Writing time to device...
    Synchronized time

If you want to regularly synchronize the time on the device, you can run this command as a service, e.g. with a systemd service or in a cron job in Linux.

These are the options that the ``set`` subcommand recognizes::

    $ bluetooth-clocks set -h
    usage: bluetooth-clocks set [-h] -a ADDRESS [-s SCAN_DURATION] [-t TIME] [-p]

    options:
      -h, --help            show this help message and exit
      -a ADDRESS, --address ADDRESS
                            Bluetooth address (e.g. 12:34:56:78:9A:BC)
      -s SCAN_DURATION, --scan-duration SCAN_DURATION
                            scan duration (default: 5 seconds)
      -t TIME, --time TIME  the time to set, in ISO 8601 format (e.g. 2023-01-10T16:20,
                            default: current time)
      -p, --am-pm           use AM/PM format (default: 24-hour format)

.. warning::

  Don't be a jerk by changing the time of other people's clocks. Use this tool responsibly.

Getting the time
----------------

Get the time from a clock with a given Bluetooth address::

    $ bluetooth-clocks get -a E7:2E:00:B1:38:96
    Scanning for device E7:2E:00:B1:38:96...
    Reading time from device...
    2023-01-14T17:54:17

These are the options that the ``get`` subcommand recognizes::

    $ bluetooth-clocks get -h
    usage: bluetooth-clocks get [-h] -a ADDRESS [-s SCAN_DURATION]

    options:
      -h, --help            show this help message and exit
      -a ADDRESS, --address ADDRESS
                            Bluetooth address (e.g. 12:34:56:78:9A:BC)
      -s SCAN_DURATION, --scan-duration SCAN_DURATION
                            scan duration (default: 5 seconds)

Usage of the library
====================

The functionality of the command-line program can also be used in your own Python programs by using this project as a library.

See the `module reference <https://bluetooth-clocks.readthedocs.io/en/latest/api/modules.html>`_ for complete API documentation.

.. inclusion-marker-before-license

Documentation
=============

Read the `online documentation <https://bluetooth-clocks.readthedocs.io>`_ of Bluetooth Clocks.

Learn more about Bluetooth Low Energy development
=================================================

If you want to learn more about Bluetooth Low Energy development, read the book `Develop your own Bluetooth Low Energy Applications for Raspberry Pi, ESP32 and nRF52 with Python, Arduino and Zephyr <https://koen.vervloesem.eu/books/develop-your-own-bluetooth-low-energy-applications/>`_ and the accompanying GitHub repository `koenvervloesem/bluetooth-low-energy-applications <https://github.com/koenvervloesem/bluetooth-low-energy-applications>`_.

License
=======

This project is provided by Koen Vervloesem as open source software with the MIT license. See the `LICENSE <https://github.com/koenvervloesem/bluetooth-clocks/blob/main/LICENSE.txt>`_ file for more information.


CHANGELOG.rst
=========
Changelog
=========

Version 0.2.0: Get time from PVVX (2023-07-04)
==============================================

This release adds support for reading the time from devices with PVVX firmware.

* Migrate code linting to Ruff, apply fixes by @koenvervloesem in https://github.com/koenvervloesem/bluetooth-clocks/pull/15
* Add MJWSD05MMC and MHO-C122 to list of supported PVVX models by @koenvervloesem in https://github.com/koenvervloesem/bluetooth-clocks/pull/16
* Assume if TYPE_CHECKING is covered by @koenvervloesem in https://github.com/koenvervloesem/bluetooth-clocks/pull/17
* Deprecate Python 3.7 by @koenvervloesem in https://github.com/koenvervloesem/bluetooth-clocks/pull/18
* Add read time command for PVVX firmware by @koenvervloesem in https://github.com/koenvervloesem/bluetooth-clocks/pull/19

Version 0.1.2: Local time, please (2023-02-02)
==============================================

This is a bugfix release. Previously the time on Qingping devices and devices running the PVVX ATC firmware was set to UTC instead of local time.

* Autoupdate pre-commit by @koenvervloesem in https://github.com/koenvervloesem/bluetooth-clocks/pull/12
* Fix local time on PVVX and Qingping devices by @koenvervloesem in https://github.com/koenvervloesem/bluetooth-clocks/pull/13

Version 0.1.1 (2023-01-23)
==========================

This is a bugfix release.

* Fix doctest: use UTC in get_time_from_bytes example by @koenvervloesem in https://github.com/koenvervloesem/bluetooth-clocks/pull/3
* Fix link to Bleak's BLEDevice in docs by @koenvervloesem in https://github.com/koenvervloesem/bluetooth-clocks/pull/4
* Add codecov badge to README by @koenvervloesem in https://github.com/koenvervloesem/bluetooth-clocks/pull/5
* Various documentation fixes by @koenvervloesem in https://github.com/koenvervloesem/bluetooth-clocks/pull/6
* Build documentation on Read The Docs with Python 3.11 by @koenvervloesem in https://github.com/koenvervloesem/bluetooth-clocks/pull/7
* Use Ubuntu 22.04/Python 3.11 for Read The Docs by @koenvervloesem in https://github.com/koenvervloesem/bluetooth-clocks/pull/8
* Update to PyScaffold v4.4 project features by @koenvervloesem in https://github.com/koenvervloesem/bluetooth-clocks/pull/9
* Import future annotations by @koenvervloesem in https://github.com/koenvervloesem/bluetooth-clocks/pull/10

Version 0.1.0 (2023-01-20)
==========================

Initial version of Bluetooth Clocks


src/bluetooth_clocks/__main__.py
"""Main entry point for Bluetooth Clocks."""
from __future__ import annotations

import asyncio
import logging
import sys
from argparse import ArgumentParser, Namespace
from datetime import datetime

from bleak.exc import BleakError

import bluetooth_clocks
from bluetooth_clocks.exceptions import TimeNotReadableError, UnsupportedDeviceError
from bluetooth_clocks.scanners import discover_clocks, find_clock

__author__ = "Koen Vervloesem"
__copyright__ = "Koen Vervloesem"
__license__ = "MIT"

_logger = logging.getLogger(__name__)


def run() -> None:
    """Calls :func:`main` passing the CLI arguments extracted from :obj:`sys.argv`.

    This function can be used as entry point to create console scripts with setuptools.
    """
    asyncio.run(main(sys.argv[1:]))


async def main(args: list[str]) -> None:
    """Wrapper allowing the bluetooth-clocks command to be called on the command line.

    This wrapper accepts string arguments.

    Args:
      args (list[str]): Command line parameters as list of strings.
    """
    cli_args = parse_args(args)
    setup_logging(cli_args.loglevel)
    await cli_args.func(cli_args)


def parse_args(args: list[str]) -> Namespace:
    """Parse command line parameters.

    Args:
      args (List[str]): command line parameters as list of strings
          (for example  ``["--help"]``).

    Returns:
      :class:`argparse.Namespace`: command line parameters namespace
    """
    parser = ArgumentParser(description="Bluetooth Clocks")
    parser.add_argument(
        "--version",
        action="version",
        version=f"bluetooth-clocks {bluetooth_clocks.__version__}",
    )
    parser.add_argument(
        "-v",
        "--verbose",
        dest="loglevel",
        help="set loglevel to INFO",
        action="store_const",
        const=logging.INFO,
    )
    parser.add_argument(
        "-vv",
        "--very-verbose",
        dest="loglevel",
        help="set loglevel to DEBUG",
        action="store_const",
        const=logging.DEBUG,
    )

    subparsers = parser.add_subparsers(title="Subcommands")
    subparsers.required = True

    # Parser for the "discover" subcommand
    parser_discover = subparsers.add_parser(
        "discover",
        help="discover supported Bluetooth clocks",
    )
    parser_discover.add_argument(
        "-s",
        "--scan-duration",
        type=float,
        default=5.0,
        help="scan duration (default: 5 seconds)",
    )
    parser_discover.set_defaults(func=discover)

    # Parser for the "get" subcommand
    parser_get = subparsers.add_parser(
        "get",
        help="get the time from a Bluetooth clock",
    )
    parser_get.add_argument(
        "-a",
        "--address",
        help="Bluetooth address (e.g. 12:34:56:78:9A:BC)",
        required=True,
    )
    parser_get.add_argument(
        "-s",
        "--scan-duration",
        type=float,
        default=5.0,
        help="scan duration (default: 5 seconds)",
    )
    parser_get.set_defaults(func=get_clock_time)

    # Parser for the "set" subcommand
    parser_set = subparsers.add_parser("set", help="set the time of a Bluetooth clock")
    parser_set.add_argument(
        "-a",
        "--address",
        help="Bluetooth address (e.g. 12:34:56:78:9A:BC)",
        required=True,
    )
    parser_set.add_argument(
        "-s",
        "--scan-duration",
        type=float,
        default=5.0,
        help="scan duration (default: 5 seconds)",
    )
    parser_set.add_argument(
        "-t",
        "--time",
        type=str,
        help="the time to set, in ISO 8601 format (e.g. 2023-01-10T16:20, default: current time)",  # noqa: E501
    )
    parser_set.add_argument(
        "-p",
        "--am-pm",
        help="use AM/PM format (default: 24-hour format)",
        action="store_true",
    )
    parser_set.set_defaults(func=set_clock_time)

    # Return complete parser
    return parser.parse_args(args)


def setup_logging(loglevel: int) -> None:
    """Setup basic logging.

    Args:
      loglevel (int): minimum loglevel for emitting messages
    """
    logformat = "[%(asctime)s] %(levelname)s:%(name)s:%(message)s"
    logging.basicConfig(
        level=loglevel,
        stream=sys.stdout,
        format=logformat,
        datefmt="%Y-%m-%d %H:%M:%S",
    )


async def discover(args: Namespace) -> None:
    """Discover Bluetooth clocks."""
    found_clock = False

    def device_found(clock: bluetooth_clocks.BluetoothClock) -> None:
        """Show details of recognized clock."""
        nonlocal found_clock
        found_clock = True
        print(f"Found a {clock.DEVICE_TYPE}: address {clock.address}", end="")
        if clock.name:
            print(f", name {clock.name}")

    print("Scanning for supported clocks...")
    await discover_clocks(device_found, args.scan_duration)

    if not found_clock:
        print("No supported clocks found")


async def get_clock_time(args: Namespace) -> None:
    """Get the time of a Bluetooth clock."""
    try:
        print(f"Scanning for device {args.address}...")
        clock = await find_clock(args.address, args.scan_duration)
        if clock:
            print("Reading time from device...")
            received_time = await clock.get_time()
            print(datetime.fromtimestamp(received_time).isoformat())
        else:
            print(f"Didn't find device {args.address}.")
    except UnsupportedDeviceError as exc:
        print(f"Unsupported device: {exc}")
    except TimeNotReadableError:
        print("Device doesn't support reading the time")
    except asyncio.exceptions.TimeoutError as exc:
        print(f"Can't connect to device {args.address}: {exc}")
    except BleakError as exc:
        print(f"Can't read from device {args.address}: {exc}")
    except AttributeError as exc:
        print(f"Can't get attribute from device {args.address}: {exc}")


async def set_clock_time(args: Namespace) -> None:
    """Set the time of a Bluetooth clock."""
    try:
        print(f"Scanning for device {args.address}...")
        clock = await find_clock(args.address, args.scan_duration)
        if clock:
            if args.time:
                timestamp = datetime.fromisoformat(args.time).timestamp()
            else:
                timestamp = None
            print("Writing time to device...")
            await clock.set_time(timestamp, args.am_pm)
            print("Synchronized time")
        else:
            print(f"Didn't find device {args.address}.")
    except UnsupportedDeviceError as exc:
        print(f"Unsupported device: {exc}")
    except asyncio.exceptions.TimeoutError as exc:
        print(f"Can't connect to device {args.address}: {exc}")
    except BleakError as exc:
        print(f"Can't write to device {args.address}: {exc}")
    except AttributeError as exc:
        print(f"Can't get attribute from device {args.address}: {exc}")


if __name__ == "__main__":
    # ^  This is a guard statement that will prevent the following code from
    #    being executed in the case someone imports this file instead of
    #    executing it as a script.
    #    https://docs.python.org/3/library/__main__.html
    run()
